ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/PDEState.java

```diff
@@ -106,85 +100,62 @@ public int compare(BaseDescription o1, BaseDescription o2) {
-						boolean p1InWorkspace = root.findContainersForLocationURI(URIUtil.toURI(p1)).length != 0;
-						IPath p2 = new Path(loc2);
-						boolean p2InWorkspace = root.findContainersForLocationURI(URIUtil.toURI(p2)).length != 0;
-						if (p1InWorkspace && !p2InWorkspace) {
-							return -1;
-						}
-						if (!p1InWorkspace && p2InWorkspace) {
-							return 1;
-						}
-					}
-				}
-			}
-		}
-		// as a last resort, compare the bundle id...
-		long id1 = s1.getBundleId();
-		long id2 = s2.getBundleId();
-		return (id1 < id2) ? -1 : ((id1 == id2) ? 0 : 1);
+	private Comparator<BaseDescription> createHigherLocalVersionFirstComparator(String systemBSN) {
+		return systemBundlesFirst(systemBSN) //
+				.thenComparing(BaseDescription::getVersion, HIGHER_VERSION_FIRST) //
+				.thenComparing(BaseDescription::getSupplier,
+						// To have a stable ordering, the version MUST be
+						// compared first, even if from different suppliers
+						Comparator.comparing(BundleDescription::getVersion, HIGHER_VERSION_FIRST)
+								.thenComparing(IN_WORKSPACE_FIRST)
+								// as a last resort, compare the bundle id...
+								.thenComparing(BundleDescription::getBundleId));
+	}
+
+	private Comparator<BaseDescription> systemBundlesFirst(String systemBSN) {
+		Function<BaseDescription, Boolean> isSystemBundle = b -> systemBSN.equals(b.getSupplier().getSymbolicName());
+		return Comparator.comparing(isSystemBundle).reversed(); // false<true
 	}
 
 	/**
 	 * Compares the given versions and prefers ".qualifier" versions over
 	 * versions with any concrete qualifier.
-	 *
-	 * @param v1
-	 *            first version
-	 * @param v2
-	 *            second version
-	 * @return a negative number, zero, or a positive number depending on if the
-	 *         first version is more desired, equal amount of desire, or less
-	 *         desired than the second version respectively
 	 */
-	private int versionCompare(Version v1, Version v2) {
-		if (v1.getMajor() == v2.getMajor() && v1.getMinor() == v2.getMinor() && v1.getMicro() == v2.getMicro()) {
-			if (v1.getQualifier().equals(v2.getQualifier())) {
-				return 0;
-			}
-			boolean q1 = "qualifier".equals(v1.getQualifier()); //$NON-NLS-1$
-			boolean q2 = "qualifier".equals(v2.getQualifier()); //$NON-NLS-1$
-			if (q1 && !q2) {
-				return -1;
-			} else if (q2 && !q1) {
-				return 1;
+	private static final Comparator<Version> HIGHER_VERSION_FIRST = Comparator //
+			.comparingInt(Version::getMajor) //
+			.thenComparingInt(Version::getMinor) //
+			.thenComparingInt(Version::getMicro) //
+			.thenComparing(PDEState::hasGenericQualifier) // false<true
+			.thenComparing(Version::getQualifier) //
+			.reversed(); //
+
+	private static final Comparator<BundleDescription> IN_WORKSPACE_FIRST = (s1, s2) -> {
+		String n1 = s1.getName();
+		String n2 = s2.getName();
+		// From previous comparison we know that both versions are fully equal
+		if (n1 != null && n1.equals(n2) && !hasGenericQualifier(s1.getVersion())) {
```

Comment:
Here is the only difference that I know of. Previously the version of the first BaseDescription was checked if it has a qualifier, but now it uses the version of the supplier for that check. I believe that makes more sense because I have never seen for example an exported package having a qualifier.
Furthermore it would make the comparator more complex.